File: modules/mesh/bus.go
package mesh

import (
	"sync"

	"github.com/google/uuid"
)

// MessageHandler is a callback invoked when a message is published to a topic.
type MessageHandler func(topic string, data interface{})

// subscription pairs a handler with its unique subscription ID.
type subscription struct {
	id      string
	handler MessageHandler
}

// MessageBus provides a simple pub/sub message bus. Subscribers register for
// a topic by name. Publish dispatches each message to all subscribers of the
// matching topic via goroutines so that handlers do not block each other.
type MessageBus struct {
	mu     sync.RWMutex
	topics map[string][]subscription
}

// NewMessageBus creates a new MessageBus ready for use.
func NewMessageBus() *MessageBus {
	return &MessageBus{
		topics: make(map[string][]subscription),
	}
}

// Subscribe registers a handler for the given topic and returns a unique
// subscription ID that can later be passed to Unsubscribe.
func (b *MessageBus) Subscribe(topic string, handler MessageHandler) string {
	id := uuid.New().String()
	b.mu.Lock()
	b.topics[topic] = append(b.topics[topic], subscription{id: id, handler: handler})
	b.mu.Unlock()
	return id
}

// Unsubscribe removes the handler identified by (topic, id). If no such
// subscription exists, Unsubscribe is a no-op.
func (b *MessageBus) Unsubscribe(topic string, id string) {
	b.mu.Lock()
	defer b.mu.Unlock()

	subs := b.topics[topic]
	for i, s := range subs {
		if s.id == id {
			b.topics[topic] = append(subs[:i], subs[i+1:]...)
			if len(b.topics[topic]) == 0 {
				delete(b.topics, topic)
			}
			return
		}
	}
}

// Publish sends data to every subscriber of the given topic. Each handler is
// invoked in its own goroutine so that slow handlers do not block others.
func (b *MessageBus) Publish(topic string, data interface{}) {
	b.mu.RLock()
	subs := make([]subscription, len(b.topics[topic]))
	copy(subs, b.topics[topic])
	b.mu.RUnlock()

	for _, s := range subs {
		go s.handler(topic, data)
	}
}

File: packages/mesh/src/bus.ts
/**
 * @module bus
 *
 * Message bus for inter-agent communication in the mesh network.
 *
 * The bus provides pub/sub messaging between agents:
 * - Topic-based message routing
 * - Multiple subscribers per topic
 * - Fire-and-forget delivery semantics
 * - Message serialization/deserialization
 *
 * The bus is the backbone of the mesh network, enabling agents
 * to communicate without direct coupling.
 */

import { v4 as uuidv4 } from "uuid";

// ---------------------------------------------------------------------------
// MessageHandler
// ---------------------------------------------------------------------------

/** Callback invoked when a message is published to a topic. */
export type MessageHandler = (topic: string, data: unknown) => void;

// ---------------------------------------------------------------------------
// Internal subscription
// ---------------------------------------------------------------------------

interface Subscription {
  id: string;
  handler: MessageHandler;
}

// ---------------------------------------------------------------------------
// MessageBus
// ---------------------------------------------------------------------------

/**
 * Simple pub/sub message bus. Subscribers register for a topic by name.
 * Publish dispatches each message to all subscribers of the matching topic
 * asynchronously so that handlers do not block each other.
 */
export class MessageBus {
  private topics = new Map<string, Subscription[]>();

  /**
   * Subscribe registers a handler for the given topic and returns a unique
   * subscription ID that can later be passed to unsubscribe.
   */
  subscribe(topic: string, handler: MessageHandler): string {
    const id = uuidv4();
    const subs = this.topics.get(topic) ?? [];
    subs.push({ id, handler });
    this.topics.set(topic, subs);
    return id;
  }

  /**
   * Unsubscribe removes the handler identified by (topic, id). If no such
   * subscription exists, unsubscribe is a no-op.
   */
  unsubscribe(topic: string, id: string): void {
    const subs = this.topics.get(topic);
    if (!subs) return;

    const idx = subs.findIndex((s) => s.id === id);
    if (idx === -1) return;

    subs.splice(idx, 1);
    if (subs.length === 0) {
      this.topics.delete(topic);
    }
  }

  /**
   * Publish sends data to every subscriber of the given topic. Each handler is
   * invoked asynchronously so that slow handlers do not block others.
   * Handler panics are caught and suppressed.
   */
  publish(topic: string, data: unknown): void {
    const subs = this.topics.get(topic);
    if (!subs) return;

    // Snapshot to avoid mutation during iteration.
    const snapshot = subs.slice();

    for (const sub of snapshot) {
      // Fire-and-forget async dispatch (mirrors goroutine behavior).
      Promise.resolve()
        .then(() => sub.handler(topic, data))
        .catch(() => {
          // Handler threw; suppress to keep bus alive.
        });
    }
  }
}
